pkg/auth/kubesecret.go

```diff
@@ -0,0 +1,182 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+var kubeSecretListener *KubeSecretListener
+
+type KubeSecretListener struct {
+	dockerConfigs map[string]*configfile.ConfigFile
+	configMu      sync.Mutex
+	informer      cache.SharedIndexInformer
+}
+
+func InitKubeSecretListener(ctx context.Context, kubeconfigPath string) {
+	if kubeSecretListener != nil {
+		return
+	}
+	kubeSecretListener = &KubeSecretListener{
+		dockerConfigs: make(map[string]*configfile.ConfigFile),
+	}
+	go func() {
+		if kubeconfigPath != "" {
+			_, err := os.Stat(kubeconfigPath)
+			if err != nil && !os.IsNotExist(err) {
+				logrus.WithError(err).Infof("kubeconfig does not exist, kubeconfigPath %s", kubeconfigPath)
+				return
+			} else if err != nil {
+				logrus.WithError(err).Infof("failed to detect kubeconfig existence, kubeconfigPath %s", kubeconfigPath)
+				return
+			}
+		}
+		loadingRule := clientcmd.NewDefaultClientConfigLoadingRules()
+		loadingRule.ExplicitPath = kubeconfigPath
+		clientConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			loadingRule,
+			&clientcmd.ConfigOverrides{},
+		).ClientConfig()
+		if err != nil {
+			logrus.WithError(err).Infof("failed to load kubeconfig")
+			return
+		}
+		clientset, err := kubernetes.NewForConfig(clientConfig)
+		if err != nil {
+			logrus.WithError(err).Infof("failed to create kubernetes client")
+			return
+		}
+		if err := kubeSecretListener.SyncKubeSecrets(ctx, clientset); err != nil {
+			logrus.WithError(err).Infof("failed to sync secrets")
+			return
+		}
+	}()
+}
+
+func (kubelistener *KubeSecretListener) addDockerConfig(key string, obj interface{}) error {
+	data, ok := obj.(*corev1.Secret).Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return fmt.Errorf("failed to get data from new object")
+	}
+	dockerConfig := configfile.ConfigFile{}
+	if err := dockerConfig.LoadFromReader(bytes.NewReader(data)); err != nil {
+		return errors.Wrap(err, "failed to load docker config json from secret")
+	}
+	kubelistener.configMu.Lock()
+	kubelistener.dockerConfigs[key] = &dockerConfig
+	kubelistener.configMu.Unlock()
+	return nil
+}
+
+func (kubelistener *KubeSecretListener) deleteDockerConfig(key string) error {
+	kubelistener.configMu.Lock()
+	delete(kubelistener.dockerConfigs, key)
+	kubelistener.configMu.Unlock()
+	_, exists := kubelistener.dockerConfigs[key]
+	if exists {
```

Comment:
The double check is not reliable due to possible insertion concurrency. I suppose we should trust Golang Map and no need to check its existence.